Setup application and non-functional requirements

anxhe opened the pull request:
# Description

## Init rails project
Initializes rails project with Docker, postgres and vagrant. 
In order to have the same environment configurations.

## Non-functional requirements
Configures rubocop , rubycritic, sandi_meter as code review tools for quality

## Setup tests
Configures rspec-rails with simplecov and simplecov-badge to measure test coverage

## Setup frontend
Configures bootstrap 5 as frontend framework with webpacker
 
 ## Issue #1 
 
 # Testing
 
Run `rails server` to test init rails app
To test connections to database:

```
vagrant up
rake db:setup
rails dbconsole
```  
To analyze code  you can run rake
  
  ## Screenshots 

![Image 2021-04-07 at 11 26 14 p m](https://user-images.githubusercontent.com/11895273/113936884-b89e5580-97f8-11eb-9324-6cb4b376277f.jpg)

anxhe committed:
Init Rails app with docker and postgres configurations

Co-authored-by: EusebioAjas <eusebioajas6@gmail.com>
Co-authored-by: RobertVel666 <robertvelasco99@gmail.com>

anxhe committed:
Add rubocop

Co-authored-by: EusebioAjas <eusebioajas6@gmail.com>
Co-authored-by: RobertVel666 <robertvelasco99@gmail.com>

anxhe committed:
Add ruby-critic, sandi_meter and Rakefile configuration

Co-authored-by: EusebioAjas <eusebioajas6@gmail.com>
Co-authored-by: RobertVel666 <robertvelasco99@gmail.com>

anxhe committed:
Add set test:

- gem rspec-rails
- gem simplecov
- gem simplecov-badge
- config simplecov in spec_helper

Co-authored-by: EusebioAjas <eusebioajas6@gmail.com>
Co-authored-by: RobertVel666 <robertvelasco99@gmail.com>

anxhe committed:
Add require in Rakefile

Co-authored-by: EusebioAjas <eusebioajas6@gmail.com>
Co-authored-by: RobertVel666 <robertvelasco99@gmail.com>

anxhe committed:
Use default postgres port

Co-authored-by: EusebioAjas <eusebioajas6@gmail.com>
Co-authored-by: RobertVel666 <robertvelasco99@gmail.com>

anxhe committed:
Setup bootstrap with webpacker

Co-authored-by: EusebioAjas <eusebioajas6@gmail.com>
Co-authored-by: RobertVel666 <robertvelasco99@gmail.com>

anxhe committed:
Fix database configuration to use username

Co-authored-by: EusebioAjas <eusebioajas6@gmail.com>
Co-authored-by: RobertVel666 <robertvelasco99@gmail.com>

RobertVel666 committed:
added password for postgres and added wdm gem, now compatible with windows

Co-authored-by: EusebioAjas <eusebioajas6@gmail.com>
Co-authored-by: Anxhe <anyelinamoreno@gmail.com>

fego-dev reviewed (dismissed):
It's fine for setup, but won't you use jquery?

RobertVel666 commented:
> It's fine for setup, but won't you use jquery?

We will not be using Jquery, we will go with Js and bootstrap_forms.

ArielMD reviewed (dismissed):
Good job, you have all the setup ready

EduardoGHdez commented on bin/setup:
```diff
@@ -0,0 +1,36 @@
+#!/usr/bin/env ruby
```
Just curious. 

Did you write this or was auto-generated?
I found it super useful

anxhe replied:
No , it was auto-gererated   😄

EduardoGHdez commented on config/credentials.yml.enc:
```diff
@@ -0,0 +1 @@
+fT7N+Lzx6KQl2EfG/PC+L8tvvg3wxw35ehpH8opveAzSS0sRJ4sCghqtgp6I7+G7Jin0RXjhPxBdKXFFl1gMAjv+HvUXlvwbmYhApvUnu+7MQTZJtjN5n2zfMB89+IGcxZYOpzbb7buTafClaWoIJIez+AVdx5tC5e5Jv2vDlq8ZwQtCpVghu0s5aebCJIsEscQ5j6RSz5LiaOKepJM8HBMPt4whfc41/c+/NCUJgIALgZv2zyrIdmo+Q5PGjROqHmLzM79p7GaCg7d1AV585kTO05ovYkfBIVCQ6f+AevBLSSUcbqTHH9jIKYHbvqnFg55mtEMikCGgx3XUtXKMIc8r9eiAqcFoCWYS8ebmuXqTuwiVhB3qjy9i19y1CeeW+EOFZxNf0Fo6dye ...
```
Take care when using the encrypted credentials

Are you going to use them or was also autogenerated?

anxhe replied:
Same as before, It was auto-generated

EduardoGHdez reviewed (commented):
I see you mentioned using Docker in this commit: https://github.com/BC-MAR-21-ROR/kata-07-rejoj-checador-equipo-03-k07/pull/2/commits/f7af2edc1ab299e96a805b020104d69e4660c46d

But can't find any setup. Did you miss adding these changes?

EduardoGHdez commented on Gemfile:
```diff
@@ -54,3 +54,5 @@ end
 # Windows does not include zoneinfo files, so bundle the tzinfo-data gem
 gem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw, :jruby]
+
+gem "rubocop", "~> 1.12"
```
I'd suggest moving this gem to the `development` group, since it's not required for any other env 😄

EduardoGHdez replied:
Also, there is a complimentary gem called `rubocop-rails`, which has some extra checks/lints for rails projects 😄 I'd suggest to take a look at it

anxhe replied:
Oh great , We will change it 👍

EduardoGHdez commented on Gemfile:
```diff
@@ -56,3 +56,9 @@ end
 
 gem "rubocop", "~> 1.12"
+
+gem "rubycritic", "~> 4.6"
+
+gem 'rubycritic-small-badge', '~> 0.2.1'
+
+gem "sandi_meter", "~> 1.2"
```
Kind of the same comment as before.

I'd suggest moving these dependencies to the development/test group. Or if it going to be used in a test env as part of your CI/CD pipeline

anxhe replied:
Fine 🙂  , We will change it too

EduardoGHdez commented on Gemfile:
```diff
@@ -62,3 +62,9 @@ gem "rubycritic", "~> 4.6"
 
 gem "sandi_meter", "~> 1.2"
+
+gem "rspec-rails", "~> 5.0"
+
+gem "simplecov", "~> 0.21.2"
+
+gem "simplecov-badge", "~> 2.0"
```
Same, but `rspec-rails` is a specific gem for testing

EduardoGHdez commented on spec/rails_helper.rb:
```diff
@@ -0,0 +1,64 @@
+# This file is copied to spec/ when you run 'rails generate rspec:install'
```
I see you opted for using `RSpec`.

Do you mind then nuking all in`test` directory?

When a new rails project is created, it comes with minitest out-of-the-box. And It'd not make sense keeping it

anxhe replied:
Yes , we will delete this

EduardoGHdez commented on Vagrantfile:
```diff
@@ -5,7 +5,7 @@ Vagrant.configure(2) do |config|
       {
        tag: "13.2-alpine",
-       map: "5433:5432", # port mapping
+       map: "5432:5432", # port mapping
```
I see what you meant when mentioning Docker in the previous commit huh.

Well, I found it interesting you want to use Vagrant from the beggining. 

I'd suggest being cautious with it.

I mean, just make sure your teammates are familiar with it, if not, let's try to keep it as simple as possible.  For some of your teammates is the first time building web-apps or even it could be their first time writing code in ruby.

Actually, I've never use Vagrant yet, and It'd be really great if you can give us a workshop about it! 🔥 

But, my point is. Let's try to focus on Ruby and RoR for this introductory kata 🥋

anxhe replied:
I agree, in fact we added it because we were having problems installing the same postgres version (13.2), as we have different operating systems. Maybe there's a better or easier way to do this, but I only knew this way and it was what I could share to help my team. I don't know much about vagrant and docker really, besides that docker was an "easy" way to use different versions of postgres and vagrant was an "easy" way to interact with docker.

Actually, only two of us use this configuration. Perhaps we can ignore Vagrantfile and config/database.yml.

EduardoGHdez commented on config/database.yml:
```diff
@@ -2,13 +2,13 @@ default: &default
   adapter: postgresql
```
In order to let users set their prefered config for the DB. I'd suggest actually not tracking this file.

EduardoGHdez commented on package.json:
```diff
@@ -6,6 +6,7 @@
     "@rails/activestorage": "^6.0.0",
     "@rails/ujs": "^6.0.0",
     "@rails/webpacker": "5.2.1",
+    "bootstrap": "^5.0.0-beta3",
```
👌🏽  cc @fego-dev FYI bootstrap 5 no longer use jQuery 🎉

EduardoGHdez commented on config/database.yml:
```diff
@@ -2,7 +2,7 @@ default: &default
   pool: <%= ENV.fetch("RAILS_MAX_THREADS") { 5 } %>
   timeout: 5000
-  user: postgres
+  username: postgres
```
That's what I mean XD

Actually, take a look at this StackOverflow question XD

https://stackoverflow.com/questions/33470898/rails-gitignore-still-tracking-my-database-yml-file

EduardoGHdez commented on .gitignore:
```diff
@@ -40,3 +40,4 @@ yarn-debug.log*
 .yarn-integrity
 .vagrant/
 coverage/
+.env
```
Actually, I'd keep it and instead ignore `.env.local`

This way, we have a common set of env-vars in `.env` and anyone can add theirs at `.env.local` which we'd be likely to ignore

anxhe replied:
Oh, good idea

EduardoGHdez commented on Gemfile:
```diff
@@ -68,3 +68,5 @@ gem "rspec-rails", "~> 5.0"
 
 gem "simplecov-badge", "~> 2.0"
+
+gem 'wdm', '>= 0.1.0' if Gem.win_platform?
```
Interesting 

TIL 🤓

EduardoGHdez reviewed (commented):
GREAT JOB 💯 

It's really good progress.

I loved you made progressive and descriptive commits. 

Left some comments 👍🏽

anxhe committed:
Change group-develoment gems:

- Move gem rubycritic
- Move gem rubycritic-small-badge
- Move gem sandimeter
- Move gem rubocop
- Add rubocop-rails

Co-authored-by: EusebioAjas <eusebioajas6@gmail.com>
Co-authored-by: RobertVel666 <robertvelasco99@gmail.com>

anxhe committed:
Move to group-test:

- gem rspec-rails
- gem simplecov
- gem simplecov-badge

Co-authored-by: EusebioAjas <eusebioajas6@gmail.com>
Co-authored-by: RobertVel666 <robertvelasco99@gmail.com>

anxhe committed:
Prettier Gemfile

Co-authored-by: EusebioAjas <eusebioajas6@gmail.com>
Co-authored-by: RobertVel666 <robertvelasco99@gmail.com>

anxhe committed:
Delete test directory we don't use mini-test

Co-authored-by: EusebioAjas <eusebioajas6@gmail.com>
Co-authored-by: RobertVel666 <robertvelasco99@gmail.com>

anxhe committed:
Add database.yml to .gitignore

Co-authored-by: EusebioAjas <eusebioajas6@gmail.com>
Co-authored-by: RobertVel666 <robertvelasco99@gmail.com>

anxhe committed:
Add Vagrantfile to .gitignore

Co-authored-by: EusebioAjas <eusebioajas6@gmail.com>
Co-authored-by: RobertVel666 <robertvelasco99@gmail.com>

anxhe committed:
Move rspec-rails to :development :test group

Co-authored-by: EusebioAjas <eusebioajas6@gmail.com>
Co-authored-by: RobertVel666 <robertvelasco99@gmail.com>

anxhe committed:
Change rubycritic configuration in Rakefile:

- Analyze both app and spec

Co-authored-by: EusebioAjas <eusebioajas6@gmail.com>
Co-authored-by: RobertVel666 <robertvelasco99@gmail.com>

marianoalarcon commented on Rakefile:
```diff
@@ -6,24 +6,27 @@ require 'rubycritic_small_badge'
+{ spec: 65, app: 90 }.each do |dir, percentage|
+  RubyCriticSmallBadge.configure do |config|
+    config.minimum_score = percentage
+  end
```
I loved that! In fact we need something similar...

SpidySamurai replied:
I'm kind of confused, where does this config come from?

anxhe replied:
it's only to not duplicate code in configuration of rubycritic in directory lib and app

marianoalarcon commented on Gemfile:
```diff
@@ -33,6 +33,7 @@ gem 'webpacker', '~> 5.0'
 group :development, :test do
   # Call 'byebug' anywhere in the code to stop execution and get a debugger console
   gem 'byebug', platforms: %i[mri mingw x64_mingw]
+  gem 'rspec-rails', '~> 5.0'
```
This seems correct to me, in fact the documentation of the gem suggests so.

marianoalarcon reviewed (approved):
I liked the atomic commits, and I think the standards they use in development are very good. In fact Rails relies on conventions over configurations, so I think it will be easier for them to adapt to the framework. Great job!
